cupy/cuda/memory.pyx

```diff
@@ -916,9 +915,12 @@ cdef class SingleDeviceMemoryPool:
 
         # find best-fit, or a smallest larger allocation
         rlock.lock_fastrlock(self._free_lock, current_thread, True)
+        # disable GC to prevent unexpected recursion. See gh-2074.
+        gc.disable()
```

Comment:
How about following code?
```
gc_enabled = gc.isenabled()
if gc.isenabled:
    gc.disable()

...

finally:
    if gc_enabled:
        gc.enable()
```

Reply:
Great idea! As more code is adding to the logic I'm thinking about using a context manager for these lock and GC manipulations. But I'm not sure if this will hamper the performance. Could you give me any advice? Thanks.